File: examples/example07_dnn_objdetect.cpp
#include "opencv2/opencv.hpp"
#include "opencv2/bioinspired.hpp"

using namespace cv;
using namespace std;

int main(int, char** argv)
{
    const int size0 = 300;
    static const char* classNames[] =
    {
        "background", "aeroplane", "bicycle", "bird", "boat",
        "bottle", "bus", "car", "cat", "chair", "cow", "diningtable",
        "dog", "horse", "motorbike", "person", "pottedplant",
        "sheep", "sofa", "train", "tvmonitor"
    };
    int maxlabel = (int)(sizeof(classNames)/sizeof(classNames[0]));

    dnn::Net net = dnn::readNetFromCaffe("MobileNetSSD_deploy_generated.prototxt",
                                  "MobileNetSSD_deploy_generated.caffemodel");
    if(net.empty())
    {
        printf("error: cannot read MobileNet SSD network\n");
        return 0;
    }

    VideoCapture cap(0);
    if( !cap.isOpened())
    {
        printf("error: cannot initialize video capture\n");
        return 0;
    }

    Mat frame, frame_r, blob;

    for(;;) {
        cap >> frame; if(frame.empty()) break;
        resize(frame, frame_r, Size(size0, size0));
        dnn::blobFromImage(frame_r, blob, 1./127.5, Size(size0, size0), 127.5);

        net.setInput(blob);
        Mat detections = net.forward();
        CV_Assert( detections.dims == 4 && (detections.size[2] == 0 || detections.size[3] == 7));
        for( int i = 0; i < detections.size[2]; i++ )
        {
            int idx[] = {0, 0, i, 0};
            float* row = detections.ptr<float>(idx);
            float confidence = row[2];
            if( confidence < 0.2f )
                continue;
            int class_id = cvRound(row[1]);
            if( class_id < 0 || class_id >= maxlabel )
                continue;
            int x0 = cvRound(row[3] * frame.cols);
            int y0 = cvRound(row[4] * frame.rows);
            int x1 = cvRound(row[5] * frame.cols);
            int y1 = cvRound(row[6] * frame.rows);
            rectangle(frame, Point(x0, y0), Point(x1, y1), Scalar(100, 255, 100));
            string label = format("%s: %.2f", classNames[class_id], confidence);
            int bl = 0;
            Size lsize = getTextSize(label, FONT_HERSHEY_SIMPLEX, 0.5, 1, &bl);
            rectangle(frame, Point(x0,y0), Point(x0+lsize.width, y0+lsize.height+bl), Scalar(100,255,100), -1);
            putText(frame, label, Point(x0, y0 + lsize.height), FONT_HERSHEY_SIMPLEX, 0.5, Scalar(0, 0, 0), 1, LINE_AA);
        }
        imshow("detections", frame);
        if( waitKey(30) >= 0 ) break;
    }
    return 0;
}

File: examples/example07_dnn_objdetect.py
import sys, numpy as np, cv2 as cv

size0 = 300
classNames = ('background', 'aeroplane', 'bicycle', 'bird', 'boat',
              'bottle', 'bus', 'car', 'cat', 'chair', 'cow', 'diningtable',
              'dog', 'horse', 'motorbike', 'person', 'pottedplant',
              'sheep', 'sofa', 'train', 'tvmonitor')
net = cv.dnn.readNetFromCaffe('MobileNetSSD_deploy_generated.prototxt',
                              'MobileNetSSD_deploy_generated.caffemodel')
cap = cv.VideoCapture(0)
while True:
    ret, frame = cap.read()
    if not ret: break
    blob = cv.dnn.blobFromImage(frame, 1./127.5, (size0, size0), 127.5)
    net.setInput(blob)
    detections = net.forward()
    for i in range(detections.shape[2]):
        confidence = detections[0, 0, i, 2]
        if confidence > 0.2:
            class_id = int(detections[0, 0, i, 1])
            x0 = int(detections[0, 0, i, 3] * frame.shape[1])
            y0 = int(detections[0, 0, i, 4] * frame.shape[0])
            x1 = int(detections[0, 0, i, 5] * frame.shape[1])
            y1 = int(detections[0, 0, i, 6] * frame.shape[0])
            cv.rectangle(frame, (x0, y0), (x1, y1), (100, 255, 100), 2)
            label = classNames[class_id] + ": " + str(round(confidence, 2))
            lsize, bl = cv.getTextSize(label, cv.FONT_HERSHEY_SIMPLEX, 0.5, 1)
            cv.rectangle(frame, (x0, y0), (x0 + lsize[0], y0 + lsize[1] + bl),
                      (100, 255, 100), cv.FILLED)
            cv.putText(frame, label, (x0, y0 + lsize[1]),
                    cv.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 0), 1, cv.LINE_AA)
    cv.imshow("detections", frame)
    if cv.waitKey(30) >= 0: break
